Migrate away from per-segment-per-threadlocals on SegmentReader

rmuir opened the pull request:
## Background

Currently the stored fields and term vectors apis on the index are "stateless".
Unlike the other parts of the APIs, users can't call any iterators/enumerators, only:
```java
indexReader.document(0);
indexReader.document(1);
// up to potentially thousands of docs because lusers do that
```

Instead of adding any real iterator, `ThreadLocal`s were added on each segment behind-the-scenes to prevent from having to `clone()` the stored fields or term vectors reader on every document. For example, this caching could reduce the amount of `NIOFSDirectory` buffer refills and other associated overhead.

But the old API from a previous time, seems to only gets worse these days, because the implementations are more complicated and do block-compression, dictionaries, etc. In some cases, the cached resources can grow to extremely large amounts (see @luyuncheng writeup in #11987 as an example)

These per-segment threadlocals can cause memory issues if you have tons of segments, tons of threads, or especially both. Seems plenty of java developers can't help but run into it.

## New API

I propose we deprecate these APIs:
* `IndexReader.document()`
* `IndexReader.termVectors()`
* `IndexSearcher.doc()`

And replace the functionality with these APIs:
* `IndexReader.storedFields()`
* `IndexReader.termVectors()`
* `IndexSearcher.storedFields()`

Instead of lucene internally caching a reader per-thread-per-segment, a user can get one themselves e.g. per-search:
```java
TopDocs hits = searcher.search(query, 10);
StoredFields storedFields = searcher.storedFields();
for (ScoreDoc hit : hits.scoreDocs) {
  Document doc = storedFields.document(hit.doc);
}
// now the StoredFields instance can be gc'd normally
```
It will re-use the datastructures across a search if someone has thousands and thousands of hits, but avoid the ThreadLocal pain.

## Deprecated API

The deprecated APIs still work the same way, and still use ThreadLocals. This way apps can safely migrate to the new APIs at their convenience. Once all the deprecations are fixed, then no ThreadLocals will be created any more, and the app can enjoy the RAM savings.

All code and tests have been moved to the new API in this PR, after backporting to 9.x, we can commit this patch to remove the deprecated apis / threadlocal support completely:  [nukeDeprecated.patch.txt](https://github.com/apache/lucene/files/10203216/nukeDeprecated.patch.txt)

rmuir committed:
Factor out user API from o.a.l.codecs.StoredFieldsReader/TermVectorsReader into o.a.l.index

These apis "StoredFields" and "TermVectors" act like the other enums on fields/postings. The codec *Readers retain all the low-level details such as checkIntegrity/clone/etc

rmuir committed:
add new api, start fixing tests

jpountz reviewed (commented):
I like the suggested direction.

For the record, another direction I've been contemplating consisted of passing a set of doc IDs to the `IndexReader#document` API, but I think it's a worse option because it doesn't really work for merging. I like this one better.

As far as fixing tests is concerned, would you be ok with fixing them in a suboptimal mechanical way, e.g. replacing every call to `indexReader.document(doc)` with `indexreader.storedFields().document(doc)`, without taking care of reusing the same `StoredFields` instance across multiple docs?

rmuir commented:
That's fine. or we could fix `newSearcher` to not wrap with crazy CodecReader's. or we could fix said CodecReaders (since they are only used for tests) to implement the deprecated document apis, like SegmentReader does. Or we could give CodecReader a default impl that isn't very performant other than UOE.

I wanted to throw the UOE, at least at first, to be sure i knew exactly what was calling old .document API (e.g. in case i forgot to fix a filter-reader). But it doesn't have to stay.

rmuir committed:
delegate the deprecated APIs in FilterCodecReader.

there's no longer a final default implementation in CodecReader (it
throws UOE), but as a practical matter, it seems to make tests happy and
isn't too invasive.

jpountz commented on lucene/test-framework/src/java/org/apache/lucene/tests/index/BaseStoredFieldsFormatTestCase.java:
```diff
@@ -813,6 +837,7 @@ public void testBulkMergeWithDeletes() throws IOException {
   }
 
   /** mix up field numbers, merge, and check that data is correct */
+  @AwaitsFix(bugUrl = "WTF with this test")
```
What is the problem? I'm not too familiar with it but it seems to test that merging correctly de-optimizes bulk merges when field numbers are not aligned, which makes sense?

jpountz replied:
I pushed a fix for this test.

rmuir replied:
thank you!

jpountz committed:
Fix mismatched test.

jpountz committed:
Add more checks on top of stored fields readers.

jpountz committed:
Fix more tests to call the proper stored fields API (mechanical change).

jpountz committed:
Fix more tests to call the proper term vectors API (mechanical change).

jpountz committed:
Fix more call sites in non-test sources.

jpountz commented:
I pushed most call sites I think. The main remaining ones are in lucene/highlighter, which require a bit more changes.

rmuir commented on lucene/queries/src/java/org/apache/lucene/queries/mlt/MoreLikeThis.java:
```diff
@@ -268,6 +270,12 @@ public final class MoreLikeThis {
   private final IndexReader ir;
 
+  /** Stored fields for {@code ir}. */
+  private final StoredFields storedFields;
```
hmm, should we do it this way as instance members? `MoreLikeThis` was thread-safe before... i think. This would now make it unsafe. Maybe we should just create these in the actual method that wants to pull from storedfields?

jpountz replied:
Good call, thanks for catching this.

jpountz committed:
Thread safety.

rmuir committed:
Fix code sample for vectors

rmuir committed:
hoist ir.storedFields() out of loops (mostly core tests)

Makes tests look more natural, perform better, and give better coverage (e.g. exercise reuse of the same instance)

rmuir committed:
Merge branch 'main' into thread_local_annihilation

rmuir committed:
fix merging bugs

rmuir committed:
Fix remaining javadocs refs to deprecated APIs in lucene/core

rmuir committed:
tidy

rmuir committed:
hoist more instances out of loops

rmuir committed:
cut over remaining stragglers

rmuir commented:
I cutover all the remaining stuff, by nuking the old api locally and making sure full gradle check passes.

attached is the patch i used to nuke the old APIs.
[nukeDeprecated.patch.txt](https://github.com/apache/lucene/files/10196690/nukeDeprecated.patch.txt)

It reveals two more things to fix:
* Fix test-framework's FieldFilterLeafReader to implement new APIs
* Look at test-framework's AssertingLeafReader to see if we can add safety checks to new APIs

rmuir committed:
implement new api on FieldFilterLeafReader

rmuir committed:
AssertingLeafReader: thread-safety checks for StoredFields,TermVectors,Terms,Fields instances

These should not be shared across threads, so add checks.

dweiss commented on lucene/test-framework/src/java/org/apache/lucene/tests/index/AssertingLeafReader.java:
```diff
@@ -113,34 +116,84 @@ public Fields getTermVectors(int docID) throws IOException {
+
   /** Wraps a Fields but with additional asserts */
   public static class AssertingFields extends FilterFields {
+    private final Thread creationThread = Thread.currentThread();
```
+1. It's not just thread-safety but strict thread-confinement. I assume we're fine with this stricter rule within tests/codebase. I don't see any reason why these objects should be passed around (?).

rmuir replied:
I had the idea to add these checks for stored/fields vectors just to fail a test clearly rather than with some "crazy" behavior that would be difficult to debug. 

Looks like i wasn't the first person to have this idea, as @jpountz already added most of them 8 years ago: https://github.com/apache/lucene/commit/fc94b0b4d9e0

For the postings API, previously we had existing checks on `TermsEnum`, `PostingsEnum`. But `Terms` and `Fields` were not checked. Seems like an oversight to me, they are all in the same codec postings api (also used by term vectors). So, depending on codec's implementation, codec might do e.g. any number of disk reads or whatever it wants and these should be checked too.

jpountz replied:
+1 to check Terms and Fields too. Most instances returned by codecs should be thread-safe but it's still good to make sure consumers don't share them across threads for consistency with other codec APIs.

rmuir committed:
docs

jpountz reviewed (approved):
It looks good. Are you good with me adding some notes on thread confinement to MIGRATE.txt and top-level javadocs of `StoredFields` and `TermVectors`?

rmuir commented:
+1 to improve docs

jpountz committed:
Add notes on thread safety.

dsmiley commented:
Love the change here to migrate away from per-segment-per-threadlocals!

When [porting this to Solr](https://github.com/apache/solr/pull/1360), I noticed a new assertion that previously didn't exist.  AssertingLeafReader now ensures that a Terms instance is only ever accessed by the thread that created it.  Terms does not document wether it's thread-safe or not but I don't believe I've ever encountered one that wasn't thread-safe.  Furthermore, the result of MultiTerms.getTerms is definitely thread-safe and can be useful to cache as there is some overhead.  Thoughts on this?

rmuir commented:
This was done on purpose as there's no such guarantee on `Terms` instances. Of particular concern are the ones returned back from `TermVectors` class. 

You can see above in the iterations on this PR that we beefed up thread-confinement tests and documentation for both `StoredFields` and `TermVectors`: these data structures were always thread-confined before by `ThreadLocal` variables. Now it is on the user and (unfortunately) it opens up the possibility of mistakes. 

We no longer "hand-hold" users with `ThreadLocal` variables, but at the same time, it allows for large memory reduction for applications that don't need them.

I don't think MultiTerms/Terms should be accessed by multiple threads either, there's no guarantee that here, that there isnt some unsafe publishing of variables updated or something like that: please don't cache them. If you insist on caching them, maybe consider using your own ThreadLocal for that purpose.

hydrogen666 commented:
In previous version `StoredFieldsReader` is cached in `ThreadLocal`, but now we need to `clone` `StoredFieldsReader` every time if we need to visit store fields. Will this PR cause any performance issue?

hydrogen666 commented:
> In previous version, `StoredFieldsReader` is cached in `ThreadLocal`, but now we need to `clone` `StoredFieldsReader` every time if we need to visit store fields. Will this PR cause any performance issue?

It is reasonable to clone `StoredFieldsReader` in one `IndexSearcher` context because one search request may hit many docs, but in some circumstances such as get by `_id` in Elasticsearch, cloning `StoredFieldsReader` every time may cause performance issue? Does my aforementioned concern make sense?

0ctopus13prime commented:
@rmuir 
Hi Robert, just curious what is the contract of calling `close` of cloned stored fields?
Codec reader gets stored fields in [here](https://github.com/apache/lucene/blob/main/lucene/core/src/java/org/apache/lucene/index/CodecReader.java#L88C33-L88C48), and SegmentReader return a cloned stored fields reader in [here](https://github.com/apache/lucene/blob/main/lucene/core/src/java/org/apache/lucene/index/SegmentReader.java#L264), but it never gets closed!

Bit confusing, but I'm assuming the contract is the shallow copied, so it's not suppose to be closed right?
Would be appreciated if you shed some lights here.
Thank you.

rmuir commented:
@0ctopus13prime see recent dev list thread on this very subject: https://lists.apache.org/thread/d14b5bftl50dmycq08kq76jwb1kk4wts

I think @uschindler summarizes it well in his last response. These slices/clones are just "views" and not backed by file descriptors or anything physical: they are effectively a `long offset` at most.

0ctopus13prime commented:
@rmuir 
Thank you for sharing! :)
